src/network/client.rs

```diff
@@ -230,13 +349,82 @@ impl NetworkClient {
     }
 
     fn handle_heartbeat(&self, peer_id: PeerId, heartbeat: Heartbeat) {
-        if !heartbeat.version_matches(&SUPPORTED_VERSIONS) {
-            metrics::IGNORED_PINGS.inc();
+        tracing::debug!("Ping from {peer_id} {:?}", heartbeat.assignment_id);
+        let local_assignments = self.assignments.lock();
+        let Some(assignment) = local_assignments.get(&heartbeat.assignment_id) else {
+            tracing::debug!("Assignment {:?} not found", heartbeat.assignment_id);
+            let latest_assignmaent_id = local_assignments
+                .last_key_value()
+                .map(|(assignment_id, _)| assignment_id.clone())
+                .unwrap_or_default();
+            if heartbeat.assignment_id > latest_assignmaent_id {
+                self.heartbeat_buffer.lock().push_back((peer_id, heartbeat));
+                tracing::debug!("Putting heartbeat into waitlist for {peer_id}");
+                {
+                    let mut heartbeat_buffer = self.heartbeat_buffer.lock();
+                    if heartbeat_buffer.len() > MAX_WAITING_PINGS {
+                        heartbeat_buffer.pop_front();
+                    }
+                }
+            } else {
+                tracing::error!("Dropping heartbeat from {peer_id}");
```

Comment:
This shouldn't be an error, it can happen occasionally. More than that, I think it shouldn't even be INFO, because this log only tells us how often such events occur, so it's better to replace it with prometheus counter instead (as done with ignored_pings)

Reply:
done